core/src/main/scala/kafka/coordinator/group/CoordinatorPartitionWriter.scala

```diff
@@ -0,0 +1,171 @@
+   */
+  override def registerListener(
+    tp: TopicPartition,
+    listener: PartitionWriter.Listener
+  ): Unit = {
+    replicaManager.maybeAddListener(tp, new ListenerAdapter(listener))
+  }
+
+  /**
+   * Deregister a PartitionWriter.Listener.
+   *
+   * @param tp       The partition to deregister the listener from.
+   * @param listener The listener.
+   */
+  override def deregisterListener(
+    tp: TopicPartition,
+    listener: PartitionWriter.Listener
+  ): Unit = {
+    replicaManager.removeListener(tp, new ListenerAdapter(listener))
+  }
+
+  /**
+   * Write records to the partitions. Records are written in one batch so
+   * atomicity is guaranteed.
+   *
+   * @param tp      The partition to write records to.
+   * @param records The list of records. The records are written in a single batch.
+   * @return The log end offset right after the written records.
+   * @throws KafkaException Any KafkaException caught during the write operation.
+   */
+  override def append(
+    tp: TopicPartition,
+    records: util.List[T]
+  ): Long = {
+    if (records.isEmpty) throw new IllegalStateException("records must be non-empty.")
+
+    replicaManager.getLogConfig(tp) match {
+      case Some(logConfig) =>
+        val magic = logConfig.recordVersion.value
+        val maxBatchSize = logConfig.maxMessageSize
+        val currentTimeMs = time.milliseconds()
+
+        val recordsBuilder = MemoryRecords.builder(
+          ByteBuffer.allocate(math.min(16384, maxBatchSize)),
+          magic,
+          compressionType,
+          TimestampType.CREATE_TIME,
+          0L,
+          maxBatchSize
+        )
+
+        records.forEach { record =>
+          val keyBytes = serializer.serializeKey(record)
+          val valueBytes = serializer.serializeValue(record)
+
+          if (recordsBuilder.hasRoomFor(currentTimeMs, keyBytes, valueBytes, EMPTY_HEADERS)) recordsBuilder.append(
+            currentTimeMs,
+            keyBytes,
+            valueBytes,
+            EMPTY_HEADERS
+          ) else throw new RecordTooLargeException(s"Message batch size is ${recordsBuilder.estimatedSizeInBytes()} bytes " +
+            s"in append to partition $tp which exceeds the maximum configured size of $maxBatchSize.")
+        }
+
+        var appendResults: Map[TopicPartition, PartitionResponse] = Map.empty
+        replicaManager.appendRecords(
+          timeout = 0L,
+          requiredAcks = 1,
+          internalTopicsAllowed = true,
+          origin = AppendOrigin.COORDINATOR,
+          entriesPerPartition = Map(tp -> recordsBuilder.build()),
+          responseCallback = results => appendResults = results,
+          // We can directly complete the purgatories here because we don't hold
+          // any conflicting locks.
+          actionQueue = directActionQueue
+        )
+
+        val partitionResult = appendResults.getOrElse(tp,
+          throw new IllegalStateException(s"Append status $appendResults should have partition $tp."))
```

Comment:
when do we expect to hit this exception?

Reply:
never, hopefully. it is only here as a defensive measure in case of a bug.